more tests for SimpleInterval.

akiezun opened the pull request:
coverage now at 100% 

fixes https://github.com/broadinstitute/gatk/issues/1405 by moving getSpanningInterval as private to ShuffleJoinReadsWithRefBases

@lbergelson can you review?

lbergelson commented:
+1 to increasing test coverage.  

I don't think moving `getSpanningInterval` is the right solution to 1405.  Why not just add the checks and leave it in SimpleInterval?  It's a useful function that will be reinvented in a number of places if it's moved out.

droazen commented:
Agree with @lbergelson on this one

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java:
```diff
@@ -97,6 +97,7 @@ public SimpleInterval(final String str){
          * So we can either keep this parsing code in the constructor or make a static factory method
          * and make multiple objects. We chose the former.
          */
+        Utils.nonNull(str);
```
We should probably add a check for empty string as well.  I don't think  the interval from 1 to max-int on conting "" is something that people would ever want.

akiezun replied:
ok

akiezun commented:
If it's useful well move it back. Right now it's used in 1 place.

On Thursday, February 18, 2016, Louis Bergelson <notifications@github.com>
wrote:

> +1 to increasing test coverage.
>
> I don't think moving getSpanningInterval is the right solution to 1405.
> Why not just add the checks and leave it in SimpleInterval? It's a useful
> function that will be reinvented in a number of places if it's moved out.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/gatk/pull/1497#issuecomment-185820758>.
>


-- 
Sent from Gmail Mobile

droazen commented:
@akiezun What will actually happen is that someone will need that functionality months from now, forget that it already exists (embedded in some random tool), and re-implement it. It should be moved back now before this is allowed to happen.

akiezun commented:
Fine but this is clearly premature optimization.

How about a class called Intervals or intervalutils for this sort of random
utility ?

On Thursday, February 18, 2016, droazen <notifications@github.com> wrote:

> @akiezun <https://github.com/akiezun> What will actually happen is that
> someone will need that functionality months from now, forget that it
> already exists (embedded in some random tool), and re-implement it. It
> should be moved back now before this is allowed to happen.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/gatk/pull/1497#issuecomment-185886728>.
>


-- 
Sent from Gmail Mobile

droazen commented:
`IntervalUtils` already exists -- I'd be fine with moving it there as a compromise. We just need to avoid the classic GATK3 mistake of embedding general utilities within tools.

akiezun commented:
done. moved getSpanningInterval to IntervalUtils and simplified + added tests for it. Moved IntervalUtilsUnitTest to the right package. No empty string in ctor.

back to @lbergelson

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java:
```diff
@@ -59,14 +59,34 @@
+        if (contigs.size() != 1){
+            throw new IllegalArgumentException("found different contigs from inputs:" + contigs);
+        }
+        final int minStart = locations.stream().mapToInt(l -> l.getStart()).min().getAsInt();
```
I think this doesn't actually fix #1405 yet.

akiezun replied:
fixed now by the call to Utils.containsNoNull

lbergelson commented:
@akiezun Add that per element null check that valentine wants and then feel free to merge.
